File: ios/Sendro/Core/NoteStore.swift
//
//  NoteStore.swift
//  Sendro
//
//  The 24-hour notes shelf (PROTOCOL.md §11.3).
//
//  A §11 text message is ephemeral ON THE WIRE and on the host: the PC keeps
//  it in RAM until it is delivered or dismissed and never writes it down.
//  That is still true and this file does not change it.
//
//  What this file adds is a LOCAL, opt-out-free convenience on the phone: a
//  copy of the text you sent or received, kept on this device only, for 24
//  hours, so you can look at that Wi-Fi password again after the card has
//  gone. The rules, deliberately narrow:
//
//  - Local only. Nothing here is ever uploaded, synced or put in history.
//  - Time-boxed. Every note carries its own expiry; anything past it is
//    deleted on load, on every write, and once a minute while the app is
//    open. There is no "keep forever" and no way to extend a note.
//  - Deletable. One swipe removes a note; "Clear all" removes the shelf.
//  - Protected at rest. The file is written with
//    `.completeFileProtection`, so it is unreadable while the phone is
//    locked, and it lives in Application Support (excluded from backups is
//    NOT needed — a 24 h file that expires on read is harmless — but it is
//    never in Documents, so it cannot show up in the Files app).
//

import Foundation
import Combine

/// One kept line of text. `Codable` here is intentional and is the ONLY
/// place in the app where message text is allowed to be serialized.
struct Note: Identifiable, Codable, Equatable {

    enum Direction: String, Codable {
        case incoming       // arrived from a paired computer
        case outgoing       // typed here and sent
    }

    let id: String
    let text: String
    /// The other end of the exchange — the PC's name.
    let peerName: String
    let direction: Direction
    let createdAt: Date
    /// Hard deletion time. Stored rather than derived so a change to the
    /// TTL can never silently extend notes that already exist.
    let expiresAt: Date

    var isExpired: Bool { Date() >= expiresAt }
}

@MainActor
final class NoteStore: ObservableObject {

    /// How long a note survives. Fixed; not a setting.
    static let ttl: TimeInterval = 24 * 60 * 60

    /// Upper bound on the shelf, oldest dropped first. Bounds the file and
    /// keeps the list readable.
    static let capacity = 200

    @Published private(set) var notes: [Note] = []

    private let fileURL: URL
    private var pruneTimer: Task<Void, Never>?

    nonisolated init(fileURL: URL? = nil) {
        self.fileURL = fileURL
            ?? AppPaths.appSupport.appendingPathComponent("notes.json", isDirectory: false)
        Task { @MainActor in self.load() }
    }

    // MARK: Lifecycle

    /// Start the once-a-minute expiry sweep. Called when the app becomes
    /// active; cheap enough to call repeatedly.
    func startPruning() {
        guard pruneTimer == nil else { return }
        pruneTimer = Task { @MainActor [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 60_000_000_000)
                if Task.isCancelled { return }
                // The store went away with the app graph — end the loop
                // rather than spinning against a dead reference.
                guard let self else { return }
                self.prune()
            }
        }
    }

    func stopPruning() {
        pruneTimer?.cancel()
        pruneTimer = nil
    }

    // MARK: Reading

    /// Newest first — the order the shelf is read in.
    var newestFirst: [Note] { notes.sorted { $0.createdAt > $1.createdAt } }

    var isEmpty: Bool { notes.isEmpty }

    /// Whole hours (rounded up, minimum 1) left before `note` is deleted.
    /// Used for the "23 h left" caption; deliberately coarse.
    static func hoursLeft(for note: Note) -> Int {
        let seconds = note.expiresAt.timeIntervalSinceNow
        guard seconds > 0 else { return 0 }
        return max(1, Int(ceil(seconds / 3600)))
    }

    // MARK: Writing

    /// Record text that arrived from a computer. Deduped by message id, so a
    /// resend of the same §11 message never stacks twice.
    func addIncoming(_ messages: [Message]) {
        guard !messages.isEmpty else { return }
        var changed = false
        for message in messages {
            let id = "in:\(message.messageId)"
            guard !notes.contains(where: { $0.id == id }) else { continue }
            let created = Date(timeIntervalSince1970: Double(message.sentAtMs) / 1000)
            // A host clock that is wrong must not create a note that is
            // already expired (or one that outlives the TTL): the shelf
            // clock is ours.
            notes.append(Note(id: id,
                              text: message.text,
                              peerName: message.senderName,
                              direction: .incoming,
                              createdAt: min(created, Date()),
                              expiresAt: Date().addingTimeInterval(Self.ttl)))
            changed = true
        }
        if changed { commit() }
    }

    /// Record text this phone sent. Called only after the host accepted it.
    func addOutgoing(_ text: String, peerName: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        notes.append(Note(id: "out:\(UUID().uuidString)",
                          text: text,
                          peerName: peerName,
                          direction: .outgoing,
                          createdAt: Date(),
                          expiresAt: Date().addingTimeInterval(Self.ttl)))
        commit()
    }

    func remove(id: String) {
        notes.removeAll { $0.id == id }
        commit()
    }

    func clearAll() {
        notes.removeAll()
        commit()
    }

    /// Drop everything past its expiry. Safe to call at any time.
    func prune() {
        let before = notes.count
        notes.removeAll { $0.isExpired }
        if notes.count != before { commit() }
    }

    // MARK: Persistence

    private func commit() {
        notes.removeAll { $0.isExpired }
        if notes.count > Self.capacity {
            notes.sort { $0.createdAt < $1.createdAt }
            notes.removeFirst(notes.count - Self.capacity)
        }
        save()
    }

    /// Read the shelf off disk. MERGES rather than replaces: `init` schedules
    /// this on a later main-actor turn, and the poll loop can land a message
    /// first — a straight assignment would silently throw that note away.
    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        guard let decoded = try? JSONDecoder().decode([Note].self, from: data) else {
            // A corrupt shelf is not worth a recovery path: it is a 24-hour
            // cache. Start clean and overwrite it on the next write.
            try? FileManager.default.removeItem(at: fileURL)
            return
        }
        let known = Set(notes.map(\.id))
        let restored = decoded.filter { !$0.isExpired && !known.contains($0.id) }
        guard !restored.isEmpty else { return }
        notes.append(contentsOf: restored)
        commit()
    }

    private func save() {
        if notes.isEmpty {
            try? FileManager.default.removeItem(at: fileURL)
            return
        }
        guard let data = try? JSONEncoder().encode(notes) else { return }
        try? data.write(to: fileURL, options: [.atomic, .completeFileProtection])
    }
}

File: android/app/src/main/java/com/sendro/android/core/NoteStore.kt
package com.sendro.android.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import java.io.File
import java.util.UUID
import kotlin.math.ceil

/**
 * The 24-hour notes shelf (PROTOCOL.md §11.3).
 *
 * A §11 text message is ephemeral ON THE WIRE and on the host: the PC keeps it
 * in RAM until it is delivered or dismissed, and never writes it down. That is
 * still true, and [MessageCenter]'s in-RAM contract is untouched.
 *
 * What this adds is a LOCAL convenience on this device: a copy of the text you
 * sent or received, kept here only, for 24 hours, so a Wi-Fi password can be
 * read again after the card is gone. The rules are deliberately narrow:
 *
 *  - Local only. Never uploaded, never synced, never in [HistoryStore].
 *  - Time-boxed. Every note carries its own expiry and is deleted on load, on
 *    every write and on every sweep. There is no way to extend one.
 *  - Deletable. One tap removes a note; "Clear" empties the shelf.
 *  - App-private. `files/state/notes.json` inside the app sandbox, which is
 *    encrypted at rest on any device with a screen lock.
 */
@Serializable
data class Note(
    val id: String,
    val text: String,
    /** The other end of the exchange — the computer's name. */
    val peerName: String,
    /** "incoming" | "outgoing" */
    val direction: String,
    val createdAtMs: Long,
    /**
     * Hard deletion time. Stored rather than derived, so changing the TTL can
     * never silently extend notes that already exist.
     */
    val expiresAtMs: Long,
) {
    val isIncoming: Boolean get() = direction == DIRECTION_IN

    fun isExpired(nowMs: Long = System.currentTimeMillis()): Boolean = nowMs >= expiresAtMs

    /** Whole hours left (rounded up, minimum 1) — the "23 h left" caption. */
    fun hoursLeft(nowMs: Long = System.currentTimeMillis()): Int {
        val remaining = expiresAtMs - nowMs
        if (remaining <= 0) return 0
        return maxOf(1, ceil(remaining / 3_600_000.0).toInt())
    }

    companion object {
        const val DIRECTION_IN = "incoming"
        const val DIRECTION_OUT = "outgoing"
    }
}

class NoteStore(private val file: File) {

    private val _notes = MutableStateFlow(load())

    /** Newest first — the order the shelf is read in. */
    val notes: StateFlow<List<Note>> = _notes.asStateFlow()

    /**
     * Record text that arrived from a computer. Deduped by message id, so a
     * resend of the same §11 message never stacks twice.
     */
    @Synchronized
    fun addIncoming(messages: List<SendroMessage>) {
        if (messages.isEmpty()) return
        val now = System.currentTimeMillis()
        val known = _notes.value.mapTo(HashSet()) { it.id }
        val fresh = messages.mapNotNull { message ->
            val id = "in:${message.messageId}"
            if (!known.add(id)) return@mapNotNull null
            Note(
                id = id,
                text = message.text,
                peerName = message.senderName,
                direction = Note.DIRECTION_IN,
                // A host clock that is wrong must not produce a note that is
                // already expired: the shelf clock is ours.
                createdAtMs = minOf(message.sentAtMs, now),
                expiresAtMs = now + TTL_MS,
            )
        }
        if (fresh.isEmpty()) return
        commit(fresh + _notes.value)
    }

    /** Record text this device sent. Called only after the host accepted it. */
    @Synchronized
    fun addOutgoing(text: String, peerName: String) {
        if (text.isBlank()) return
        val now = System.currentTimeMillis()
        val note = Note(
            id = "out:${UUID.randomUUID()}",
            text = text,
            peerName = peerName,
            direction = Note.DIRECTION_OUT,
            createdAtMs = now,
            expiresAtMs = now + TTL_MS,
        )
        commit(listOf(note) + _notes.value)
    }

    @Synchronized
    fun remove(id: String) = commit(_notes.value.filterNot { it.id == id })

    @Synchronized
    fun clearAll() = commit(emptyList())

    /** Drop everything past its expiry. Safe to call at any time. */
    @Synchronized
    fun prune() {
        val before = _notes.value
        val after = before.filterNot { it.isExpired() }
        if (after.size != before.size) commit(after)
    }

    private fun commit(list: List<Note>) {
        val next = list
            .filterNot { it.isExpired() }
            .sortedByDescending { it.createdAtMs }
            .take(CAPACITY)
        _notes.value = next
        save(next)
    }

    private fun load(): List<Note> {
        if (!file.isFile) return emptyList()
        val decoded = runCatching { SendroJson.decodeFromString<List<Note>>(file.readText()) }
            // A corrupt shelf is not worth a recovery path: it is a 24-hour
            // cache. Start clean and overwrite it on the next write.
            .getOrDefault(emptyList())
        return decoded.filterNot { it.isExpired() }.sortedByDescending { it.createdAtMs }
    }

    private fun save(list: List<Note>) {
        runCatching {
            if (list.isEmpty()) {
                file.delete()
                return@runCatching
            }
            file.parentFile?.mkdirs()
            val temp = File(file.parentFile, file.name + ".tmp")
            temp.writeText(SendroJson.encodeToString(list))
            if (!temp.renameTo(file)) {
                file.writeText(temp.readText())
                temp.delete()
            }
        }
    }

    companion object {
        /** How long a note survives. Fixed; not a setting. */
        const val TTL_MS = 24 * 60 * 60 * 1000L

        /** Upper bound on the shelf, oldest dropped first. */
        const val CAPACITY = 200
    }
}
